### PR TITLE
GPT-OSS: Add user guide and tests

### PR DESCRIPTION
**Pull request by shuningjin**
# Description

Main change: for gpt-oss
- add user guide: 
  - `end_to_end/tpu/gpt_oss/run_gpt_oss.md`
  - use gpt-oss-20b on v5p-8 (sparse_matmul + megablox) as example
- add test script
  - `end_to_end/tpu/gpt_oss/20b/test_gpt_oss.sh`
  - `end_to_end/tpu/gpt_oss/120b/test_gpt_oss.sh`
- pyconfig validation

Other change: format output from forward_logit_checker, correct typos in other files

# Tests

single script for tpu: checkpoint conversion, logit check, train, sft, decode (instead of splitting conversion to another)

gpt-oss-20b, end-to-end test on v5p-8 cluster, takes around 45min
- cmd: https://paste.googleplex.com/5491652676419584
- log: https://cloudlogging.app.goo.gl/DzrB6DY1qw576Q747
- to faciliate the test, I have uploaded the dequantized checkpoint and golden logits to GCS
- note: gpt-oss with flash attention requires jax>=0.7.2

The test will also be onboarded to XLML: https://github.com/GoogleCloudPlatform/ml-auto-solutions/pull/913

## Additional Notes

HF tokenizer:
- Here we directly use the HF tokenizer, instead of storing it as assets in the repo. This practice seems fine, as long as the HF access token is handled correctly.

HF access token:
- For MaxText native training, the access token is not needed because the HF model/tokenizer (gpt-oss) and HF dataset (used in SFT) are not gated. 
- However, decoding with HF tokenizer still needs access token despite that the model is not gated. The tokenizer is built from JetStream utils, which handles things differently from MaxText.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: end_to_end/tpu/deepseek/Run_DeepSeek.md [hunk 1]
```diff
@@ -131,7 +131,7 @@ python3 -m MaxText.sft_trainer src/MaxText/configs/sft.yml \
 ```
 
 ## Decoding
-One example command to run supervised finetuning with V3 on v5p-256 with unscanned checkpoint for fast decoding.
+One example command to run decoding with V3 on v5p-256 with unscanned checkpoint for fast decoding.
```
Comment (RissyRan) [inline end_to_end/tpu/deepseek/Run_DeepSeek.md:134]: Thanks!

File: end_to_end/tpu/gpt_oss/20b/test_gpt_oss.sh [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+# default golden_logits_path=/deps/src/MaxText/test_assets/golden_data_{MODEL_NAME}.jsonl, copied from gs://maxtext-test-assets/golden_data_${MODEL_NAME}.jsonl
+python3 -m tests.forward_pass_logit_checker "${MAXTEXT_PKG_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/MaxText}/"configs/base.yml base_output_directory=${BASE_OUTPUT_PATH} run_name=forward_logits_check model_name=${MODEL_NAME} load_parameters_path=${UNSCANNED_CKPT_PATH} scan_layers=false attention=dot_product sparse_matmul=True megablox=True per_device_batch_size=1 max_target_length=4 max_prefill_predict_length=4 dtype=float32 --atol=0.1 --rtol=0.1 --max_kl_div=3e-4
+
+# Run pre-training - megablox implementation
+python3 -m MaxText.train "${MAXTEXT_PKG_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/MaxText}/"configs/base.yml base_output_directory=${BASE_OUTPUT_PATH} run_name=megablox_pre_training model_name=${MODEL_NAME} tokenizer_type=huggingface tokenizer_path=${TOKENIZER_PATH} dataset_type=synthetic enable_checkpointing=false attention=flash sparse_matmul=True megablox=True dtype=bfloat16 weight_dtype=bfloat16 per_device_batch_size=4 steps=5 max_target_length=1024 ici_fsdp_parallelism=4
+
+# Run fine-tuning - megablox implementation
```
Comment (RissyRan) [inline end_to_end/tpu/gpt_oss/20b/test_gpt_oss.sh:59]: extra comments?

**Review by RissyRan — COMMENTED** [review]
Could you also add announcement to the Main page to claim the support for gpt-oss models?

File: end_to_end/tpu/gpt_oss/20b/test_gpt_oss.sh (hunk 2 above)
Comment (shuningjin) [reply end_to_end/tpu/gpt_oss/20b/test_gpt_oss.sh:59]: Finetuning and maxtext-SFT are very similar tasks. Do we need test both? On the other hand, seems we are switching to Tunix-SFT in the near future. I am torn.

Comment (shuningjin) [reply end_to_end/tpu/gpt_oss/20b/test_gpt_oss.sh:59]: Added back finetuning.

**Comment by shuningjin** [comment]
> Could you also add announcement to the Main page to claim the support for gpt-oss models?

I just updated "README.md" and "supported_models_and_architectures.md".

File: end_to_end/tpu/gpt_oss/20b/test_gpt_oss.sh (hunk 2 above)
Comment (RissyRan) [reply end_to_end/tpu/gpt_oss/20b/test_gpt_oss.sh:59]: I am also fine actually, no strong option. But I see you here commented out, but not in 120b test script (if I recall correctly)?

File: end_to_end/tpu/gpt_oss/120b/test_gpt_oss.sh [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+# Run pre-training - megablox implementation
+python3 -m MaxText.train "${MAXTEXT_PKG_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/MaxText}/"configs/base.yml base_output_directory=${BASE_OUTPUT_PATH} run_name=megablox_pre_training model_name=${MODEL_NAME} tokenizer_type=huggingface tokenizer_path=${TOKENIZER_PATH} dataset_type=synthetic enable_checkpointing=false attention=flash sparse_matmul=True megablox=True dtype=bfloat16 weight_dtype=bfloat16 per_device_batch_size=4 steps=5 max_target_length=1024 ici_fsdp_parallelism=32
+
+# Run fine-tuning - megablox implementation
+python3 -m MaxText.train "${MAXTEXT_PKG_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/MaxText}/"configs/base.ymlbase_output_directory=${BASE_OUTPUT_PATH} run_name=megablox_fine_tuning model_name=${MODEL_NAME}tokenizer_type=huggingface tokenizer_path=${TOKENIZER_PATH} dataset_path=${DATASET_PATH}enable_checkpointing=true async_checkpointing=false load_parameters_path=${SCANNED_CKPT_PATH} scan_layers=Trueattention=flash sparse_matmul=True megablox=True dtype=bfloat16 weight_dtype=bfloat16per_device_batch_size=4 steps=5 max_target_length=1024 ici_fsdp_parallelism=1 ici_expert_parallelism=32
+
```
[the remark this answers is not in the document]
Comment (shuningjin) [reply end_to_end/tpu/gpt_oss/120b/test_gpt_oss.sh:61]: Good catch on the missing spaces! I have updated.

File: end_to_end/tpu/gpt_oss/20b/test_gpt_oss.sh [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+python3 -m MaxText.train "${MAXTEXT_PKG_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/MaxText}/"configs/base.ymlbase_output_directory=${BASE_OUTPUT_PATH} run_name=megablox_fine_tuning model_name=${MODEL_NAME}tokenizer_type=huggingface tokenizer_path=${TOKENIZER_PATH} dataset_path=${DATASET_PATH}enable_checkpointing=true async_checkpointing=false load_parameters_path=${SCANNED_CKPT_PATH} scan_layers=Trueattention=flash sparse_matmul=True megablox=True dtype=bfloat16 weight_dtype=bfloat16per_device_batch_size=4 steps=5 max_target_length=1024 ici_fsdp_parallelism=1 ici_expert_parallelism=4
+
```
[the remark this answers is not in the document]
Comment (shuningjin) [reply end_to_end/tpu/gpt_oss/20b/test_gpt_oss.sh:61]: Good catch on the missing spaces! I have updated.

File: docs/reference/supported_models_and_architectures.md [hunk 5]
```diff
@@ -49,6 +49,11 @@ MaxText is an open-source, high-performance LLM framework written in Python/JAX.
 * **Variants**: Dense (0.6B–32B); MoE (30B-A3B, 235B-A22B, 480B Coder)
 * **Notes**: **QK-Norm**, GQA, SwiGLU, RMSNorm, RoPE.
 
+### GPT-OSS
+
+* **Variants**: 20B, 120B
+* **Notes**: Local–Global interleaved attention, GQA, attention sink; YaRN-style scaling; MoE.
```
Comment (parambole) [inline docs/reference/supported_models_and_architectures.md:55]: What do the notes over here signify ?

File: end_to_end/tpu/deepseek/Run_DeepSeek.md [hunk 6]
```diff
@@ -168,7 +168,7 @@ One example command to generate golden logits from HuggingFace for V2-Lite.
 python3 -m MaxText.scratch_code.generate_hf_golden_logits \
     --model-id=deepseek-ai/DeepSeek-V2-Lite \
     --output-path=golden_DeepSeek-V2-Lite.jsonl \
-    --prompts='I love to,Today is a,What is the'
+    --prompts='I love to;Today is a;What is the'
```
Comment (parambole) [inline end_to_end/tpu/deepseek/Run_DeepSeek.md:171]: @shuningjin can you please elaborate on this change?

File: end_to_end/tpu/gpt_oss/120b/test_gpt_oss.sh [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+# Non-Googlers please remember to point `CKPT_BUCKET` to GCS buckets that you own
+# Copying the HF checkpoint into a local directory `/tmp` -- you are free to use a different directory
+if [ -z "${CKPT_DISK_LOCATION}" ]; then
+  export CKPT_BUCKET=gs://maxtext-model-checkpoints/gpt-oss-120b/hf-bf16
+  gcloud storage cp -r ${CKPT_BUCKET} /tmp
+  export CKPT_DISK_LOCATION=/tmp/hf-bf16
+fi
+
+# 1.1 Convert checkpoint to `scanned` format, more suitable for training 
+JAX_PLATFORMS=cpu python3 -m MaxText.convert_gpt_oss_ckpt --base-model-path ${CKPT_DISK_LOCATION} --maxtext-model-path ${BASE_OUTPUT_PATH}/scanned --model-size ${MODEL_NAME}
```
Comment (parambole) [inline end_to_end/tpu/gpt_oss/120b/test_gpt_oss.sh:40]: Are we also planning to onboard it to the new checkpoint conversion util?

File: end_to_end/tpu/gpt_oss/run_gpt_oss.md [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+    attention=flash \
+    sparse_matmul=True \
+    megablox=True \
+    dtype=bfloat16 \
+    weight_dtype=bfloat16 \
+    per_device_batch_size=4 \
+    steps=5 \
+    max_target_length=1024 \
+    ici_fsdp_parallelism=1 \
+    ici_expert_parallelism=4
```
Comment (parambole) [inline end_to_end/tpu/gpt_oss/run_gpt_oss.md:107]: Curious here and everywhere else why for training FSDP is set and for fine tuning expert parallelism. Can you please elaborate ?

File: end_to_end/tpu/deepseek/Run_DeepSeek.md (hunk 6 above)
Comment (shuningjin) [reply end_to_end/tpu/deepseek/Run_DeepSeek.md:171]: The prompts are now split by ";", see [line](https://github.com/AI-Hypercomputer/maxtext/blob/fafdeaa14183a8f5ca7b9f7b7542ce1655237574/src/MaxText/scratch_code/generate_hf_golden_logits.py#L156)

Previously, the delimiter was ",". The change was introduced recently in this [commit](https://github.com/AI-Hypercomputer/maxtext/commit/5d5487bfcc49c872a9fd76087335ba3e1846bfb1).

File: docs/reference/supported_models_and_architectures.md (hunk 5 above)
Comment (shuningjin) [reply docs/reference/supported_models_and_architectures.md:55]: From the [context](https://screenshot.googleplex.com/3Sz6XMyUUiGkcQp), I think "notes" means architecture features of a model family.

This page was recently added by Gagik, in [PR 2273](https://github.com/AI-Hypercomputer/maxtext/pull/2273). I am including GPT-OSS here to keep it up-to-date.

File: end_to_end/tpu/gpt_oss/120b/test_gpt_oss.sh (hunk 7 above)
Comment (shuningjin) [reply end_to_end/tpu/gpt_oss/120b/test_gpt_oss.sh:40]: We could onboard it later on, maybe as P1 or P2 task. @RissyRan 

From my initial attempt, the util needs some non-trivial changes (even for 20b)
- for unscanned conversion, the transformation only supports 1-1 mapping, whereas here we need 1-N mapping (e.g., weight X in hf is split to Y and Z in maxtext).
- for scanned conversion, we need additionally support inhomogenous scannable block (gpt-oss, llama4, gemma3).  The layers are not all the same (e.g., interleaved attention pattern), so cannot simply stack them together.

File: end_to_end/tpu/gpt_oss/run_gpt_oss.md (hunk 8 above)
Comment (shuningjin) [reply end_to_end/tpu/gpt_oss/run_gpt_oss.md:107]: I am copying from [Run_DeepSeek](https://screenshot.googleplex.com/8vsDpPZMJSXsN9e). In that page, example show FSDP for training, EP for finetuning/sft, and TP for decoding.

I followed these practices for GPT-OSS. I think the goal is to showcase different parallelism that we commonly use for MoE.

File: docs/reference/supported_models_and_architectures.md (hunk 5 above)
Comment (parambole) [reply docs/reference/supported_models_and_architectures.md:55]: > From the [context](https://screenshot.googleplex.com/3Sz6XMyUUiGkcQp), I think "notes" means architecture features of a model family.

Thanks for elaborating. So are these new features or features that we want to highlight for a particular model ?

File: end_to_end/tpu/deepseek/Run_DeepSeek.md (hunk 6 above)
Comment (parambole) [reply end_to_end/tpu/deepseek/Run_DeepSeek.md:171]: Awesome! Thanks for updating.

File: end_to_end/tpu/gpt_oss/120b/test_gpt_oss.sh (hunk 7 above)
Comment (parambole) [reply end_to_end/tpu/gpt_oss/120b/test_gpt_oss.sh:40]: These are great points. Can you please create a bug for tracking it? as these features will help all future models including Qwen3-Next.

cc: @Rohan-Bierneni

File: end_to_end/tpu/gpt_oss/run_gpt_oss.md (hunk 8 above)
Comment (parambole) [reply end_to_end/tpu/gpt_oss/run_gpt_oss.md:107]: I see. 

nit: probably we can add some notes around these cause from reading the flow it looks like we are recommending different Sharding types for different use cases. 

cc: @RissyRan

File: docs/reference/supported_models_and_architectures.md (hunk 5 above)
Comment (shuningjin) [reply docs/reference/supported_models_and_architectures.md:55]: My understanding is that these are not necessarily novel features, but **fact sheet about the key component** of models. 

-> The components I listed are those I have high confidence in. I may update with a follow-up PR.

File: end_to_end/tpu/gpt_oss/120b/test_gpt_oss.sh (hunk 7 above)
Comment (shuningjin) [reply end_to_end/tpu/gpt_oss/120b/test_gpt_oss.sh:40]: -> Yes, I will create a bug to track and cc you later today.

File: end_to_end/tpu/gpt_oss/run_gpt_oss.md (hunk 8 above)
Comment (shuningjin) [reply end_to_end/tpu/gpt_oss/run_gpt_oss.md:107]: Thanks for the suggestion. 

-> I will double check, and may update with a follow-up PR.
